src/book/03-guides/01-main-codebase/03-recording-playback.mdx

```diff
@@ -0,0 +1,120 @@
+```
+
+### Stats
+
+Decodes an NBS file into a series of file statistics.
+
+| Argument            | Definition                                                        |
+| ------------------- | ----------------------------------------------------------------- |
+| --message-timestamp | Uses the message's `.timestamp` field as the time of the message. |
+
+### JSON
+
+Outputs the data in the NBS file to a JSON file. It is useful to redirect the output to a file by appending `> file.json` to the command.
+
+### Extract Images
+
+Decodes an NBS file and extracts any compressed jpeg files into jpeg files.
+
+| Argument | Definition                                  | Default                       |
+| -------- | ------------------------------------------- | ----------------------------- |
+| --output | Specifies the folder to save the images to. | The current working directory |
+
+### Video
+
+This will take images in an NBS file and create a video.
+
+| Argument  | Definition                                            | Default                       |
+| --------- | ----------------------------------------------------- | ----------------------------- |
+| --output  | Specifies the folder to save the video to.            | The current working directory |
+| --quality | Specifies the quality to save the video in.           | 30M                           |
+| --encoder | specifies the encoder to use when encoding the video. | h264_nvenc                    |
+
+### Calibrate Cameras
+
+Calculate camera intrinsics using an NBS recording of asymmetric circles grid patterns.
+
+| Argument        | Definition                                                           | Default  |
+| --------------- | -------------------------------------------------------------------- | -------- |
+| --config        | The path to the directory containing the camera configuration files. | Required |
+| --rows          | The number of rows in the asymmetric circles grid.                   | 4        |
+| --cols          | The number of columns in the asymmetric circles grid.                | 11       |
+| --grid_size     | The distance between rows/cols in the grid in meters.                | 0.04     |
+| -- no-intrisics | Does not change the intrinsic configuration values.                  | -        |
+| --no-extrinsics | Does not change the extrinsic configuration values.                  | -        |
+
+### Foot Down
+
+Trains a foot down network using sensor data from the legs.
+
+## Playing Back Data
+
+Messages can be emitted from an NBS file. After recording data in an NBS file, you can play it back to the system. Following on from the previous example, lets now run `sensortest.role` with recorded sensor data to analyse the results more closely off the robot.
+
+1. Remove any modules that emit the message you recorded. In this example, remove the Sensors message. That is, remove `platform::darwin::SensorFilter` from the role.
```

Comment:
```suggestion
1. Remove any modules that emit the message you recorded. In this example, remove the SensorFilter module, since it emits `Sensors`. To do so, remove `platform::darwin::SensorFilter` from the role.
```